sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformation.scala

```diff
@@ -53,28 +78,176 @@ case class ScriptTransformation(
+            throw new NoSuchElementException
+          }
+ 
+          if (outputSerde == null) {
+            val prevLine = curLine
             curLine = reader.readLine()
+ 
+            if (!ioschema.schemaLess) {
+              new GenericRow(
+                prevLine.split(outputRowFormatMap("TOK_TABLEROWFORMATFIELD"))
+                .asInstanceOf[Array[Any]])
+            } else {
+              new GenericRow(
+                prevLine.split(outputRowFormatMap("TOK_TABLEROWFORMATFIELD"), 2)
+                .asInstanceOf[Array[Any]])
+            }
+          } else {
+            val ret = deserialize()
+            if (!eof) {
+              cacheRow = null
+              cacheRow = deserialize()
+            }
+            ret
           }
         }
       }
-      readerThread.start()
+
+      val inputSerde: AbstractSerDe = if (ioschema.inputSerdeClass != "") {
+        val trimed_class = ioschema.inputSerdeClass.split("'")(1)
+        Utils.classForName(trimed_class)
+          .newInstance.asInstanceOf[AbstractSerDe]
+      } else {
+        null
+      }
+ 
+      if (inputSerde != null) {
+        val columns = input.map { case e: NamedExpression => e.name }.mkString(",")
+        val columnTypes = input.map { case e: NamedExpression =>
+          e.dataType.toTypeInfo.getTypeName() 
+        }.mkString(",")
+ 
+        var propsMap = ioschema.inputSerdeProps.map(kv => {
+          (kv._1.split("'")(1), kv._2.split("'")(1))
+        }).toMap + (serdeConstants.LIST_COLUMNS -> columns)
+        propsMap = propsMap + (serdeConstants.LIST_COLUMN_TYPES -> columnTypes)
+
+        val properties = new Properties()
+        properties.putAll(propsMap)
+        inputSerde.initialize(null, properties)
+      }
+      
+      val inputSoi = if (inputSerde != null) {
+        val fieldNames = input.map { case e: NamedExpression => e.name }
+        val fieldObjectInspectors = input.map { case e: NamedExpression =>
+          toInspector(e.dataType)
+        }
+        ObjectInspectorFactory
+          .getStandardStructObjectInspector(fieldNames, fieldObjectInspectors)
+          .asInstanceOf[ObjectInspector]
+      } else {
+        null
+      }
+
+      val dataOutputStream = new DataOutputStream(outputStream)
       val outputProjection = new InterpretedProjection(input, child.output)
       iter
         .map(outputProjection)
-        // TODO: Use SerDe
-        .map(_.mkString("", "\t", "\n").getBytes("utf-8")).foreach(outputStream.write)
+        .foreach { row =>
+          if (inputSerde == null) {
+            val data = row.mkString("", inputRowFormatMap("TOK_TABLEROWFORMATFIELD"),
+            inputRowFormatMap("TOK_TABLEROWFORMATLINES")).getBytes("utf-8")
+ 
+            outputStream.write(data)
+          } else {
+            val writable = inputSerde.serialize(row.asInstanceOf[GenericRow].values, inputSoi)
+            if (writable.isInstanceOf[AvroGenericRecordWritable] && HiveShim.version == "0.13.1") {
```

Comment:
Put the `if` into the `HiveShim`? BTW, I am not so sure why we need to handle the `Avro` as a specific case, I think we'd better keep it as generic purpose for this fixing. (We may not able to enumerate all of the specific case)